src/lib/io/include/sol-uart.h

```diff
@@ -238,7 +251,18 @@ void sol_uart_close(struct sol_uart *uart);
  * @note Caller should guarantee that tx buffer will not be freed until
  * callback is called.
  */
-bool sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
+int sol_uart_write(struct sol_uart *uart, const uint8_t *tx, size_t length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
+
+/**
+ * @brief Read from UART bus
+ *
+ * @param uart The UART bus handle
+ * @param rx Where to store the UART data
+ * @param length number of bytes to read
+ *
+ * @return the number of bytes read or negative errno on error.
+ */
+int sol_uart_read(struct sol_uart *uart, uint8_t *rx, size_t length);
```

Comment:
sol_buffer! and document the flags, you may take an already initialized buffer, if fixed capacity you read up to that, if it's not, then you can allocate to return all that is available.